[oneDAL] Data management section (#141)
oneapi-src/oneAPI-spec: source/elements/oneDAL/source/data_management/accessors.rst

.. _accessors:

=========
Accessors
=========

> This section contain definisions of classes that implement :ref:`accessor`

Reviewer: ```suggestion
This section contains definitions of classes that implement the :ref:`accessor`
```

concept.

------------------------------

> Requirements on accessor types

Reviewer: maybe just requirements?

------------------------------

> Each implementation of :ref:`accessor` concept in |dal_short_name| shall follow
> next requirements:

Reviewer: ```suggestion
Each implementation of the :ref:`accessor` concept in |dal_short_name| shall follow
these requirements:
```

> 1. Define single :term:`format of the data <Data format>` of the
>    accessor. Every single accessor type shall provide only one :term:`format of
>    the data <Data format>` returned from it or put inside it.

Reviewer: ```suggestion
1. Define a single :term:`format of the data <Data format>` for the
   accessor. Every single accessor type shall return and use
   only one :term:`format of the data <Data format>` .
```
`use` might not be the correct word, but `put inside it` is unclear
why not use just `data format` instead of the seonc `format of the data`?

> 2. Allow to make access to at least of one in-memory :ref:`dataset`

Reviewer 2: provide access ..

> 2. Allow to make access to at least of one in-memory :ref:`dataset`
>    implementation (like :code:`table` or its sub-types or :ref:`table builders
>    <table-builder>`).

Reviewer: ```suggestion
2. Allow it to have access to at least of one in-memory :ref:`dataset`
   implementation (such as :code:`table`, its sub-types, or :ref:`table builders
   <table-builder>`).
```
allow an accessor? I think it should work without `allow` at all: Have access to...

> 3. Allow to make read-only, write-only and read-write access to the data. If

Reviewer 2: provide read-only...

   accessor supports several :ref:`dataset` implementations to be passed in, it

>    is not necessary to support all access modes for every input object. For

Reviewer 2: wording is complex. Consider re-wording after 0.8

   example, tables shall support single read-only mode according to their

>    :ref:`concept <table>` definision.

Reviewer: ```suggestion
3. Allow it to have read-only, write-only, or read-write access to the data. If an
   accessor supports several :ref:`dataset` implementations to be passed in, it
   is not necessary for an accessor to support all access modes for every input object. For
   example, tables shall support a single read-only mode according to their
   :ref:`concept <table>` definition.
```

> 3. Allow to make read-only, write-only and read-write access to the data. If
>    accessor supports several :ref:`dataset` implementations to be passed in, it
>    is not necessary to support all access modes for every input object. For
>    example, tables shall support single read-only mode according to their
>    :ref:`concept <table>` definition.

Reviewer: ```suggestion
3. Allow it to have read-only, write-only, or read-write access to the data. If an
   accessor supports several :ref:`dataset` implementations to be passed in, it
   is not necessary for an accessor to support all access modes for every input object. For
   example, tables shall support a single read-only mode according to their
   :ref:`concept <table>` definition.
```

`use` might not be the correct word, but `put inside it` is unclear
why not use just `data format` instead of the second `format of the data`?

> 4. Shall name reading and writing operations consistently. Reading operation

Reviewer 2: provide the names for read and write operations following the pattern:
- pull_  for reading
- push_ for..

> 4. Shall name reading and writing operations consistently. Reading operation
>    name shall follow next pattern: :code:`pull_*()`, while writing operation
>    shall be named with accordance to following: :code:`push_*()`.

Reviewer: ```suggestion
4. Shall name reading and writing operations consistently. Their names 
   shall follow these patterns: :code:`pull_*()` for a reading operation
   and :code:`push_*()` for a writing operation.
```

>    shall be named with accordance to folllowing: :code:`push_*()`.

Reviewer: ```suggestion
4. Shall name reading and writing operations consistently. The names for reading 
   and writing operations shall follow  :code:`pull_*()` and :code:`push_*()` respectively.
```

```suggestion
4. Shall name reading and writing operations consistently. Their names 
   shall follow these patterns: :code:`pull_*()` for a reading operation
   and :code:`push_*()` for a writing operation.
```

5. Shall be lightweight. Its constructors from :ref:`dataset` implementations

>    shall not have heavy operations like copy of data, reading, writing, any sort

Reviewer: ```suggestion
   shall not have heavy operations such as copy of data, reading, writing, any sort
```

   of conversions. These operations shall be performed by heavy operations
   :code:`pull_*()` and :code:`push_*()`. It is not necessary to have copy- or
   move- constructors for accessor implementations since it shall be designed

>    for use in the local scope.

Reviewer: ```suggestion
   for use in a local scope.
```

----------------------------------

> Accessor types in |dal_short_name|

Reviewer: without oneDAL maybe?

----------------------------------

> |dal_short_name| defines a set of accessor classes each associated with single

Reviewer 2: duplication

> |dal_short_name| defines a set of accessor classes each associated with single
> way how to interact with data inside a :ref:`dataset`. Next table briefly
> explains these classes and shows which :ref:`dataset` implementations are
> supported by each accessor.

Reviewer: ```suggestion
|dal_short_name| defines a set of accessor classes. Each class is associated
with a single specific way of interacting with data within a :ref:`dataset`. The following table briefly
explains these classes and shows which :ref:`dataset` implementations are
supported by each accessor type.
```

.. list-table::
   :header-rows: 1
   :widths: 10 45 45

   * - Accessor type
     - Description
     - List of supported types
   * - row_accessor_

>      - Provides an access to the range of dataset rows as one :term:`contiguous

Reviewer: ```suggestion
     - Provides access to the range of dataset rows as one :term:`contiguous
```

>        <Contiguous data>` and :term:`homogeneous <Homogeneous data>` block of memory.

Reviewer 2: remove and in both rows of the table

     - :code:`homogen_table`, :code:`soa_table`, :code:`aos_table`,
       :code:`csr_table`, and their builders.
   * - column_accessor_

>      - Provides an access to the values range of a single column as one

Reviewer: ```suggestion
     - Provides access to the range of values within a single column as one
```

.. _accessors:

=========
Accessors
=========

This section defines `requirements <accessor_reqs_>`_ to an :ref:`accessor
<accessor>` implementation and introduces several `accessor types
<accessor_types_>`_.

.. _accessor_reqs:

------------
Requirements
------------

An implementation of each accessor shall satisfy the following requirements:

1. Define a single :term:`format of the data <Data format>` for the
   accessor. Every single accessor type shall return and use only one data
   format.

> 2. Allow it to have access to at least of one in-memory :ref:`dataset`

Reviewer: ```suggestion
2. Allow it to have access to at least one in-memory :ref:`dataset`
```
